sources/ElkArte/Graphics/AbstractManipulator.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+/**
+ * This file deals with low-level graphics operations performed on images,
+ * specially as needed for avatars (uploaded avatars), attachments, or
+ * visual verification images.
+ *
+ * @package   ElkArte Forum
+ * @copyright ElkArte Forum contributors
+ * @license   BSD http://opensource.org/licenses/BSD-3-Clause (see accompanying LICENSE.txt file)
+ *
+ * @version 2.0 dev
+ *
+ */
+
+namespace ElkArte\Graphics;
+
+/**
+ * Class AbstractManipulator
+ *
+ * @package ElkArte\Graphics
+ */
+abstract class AbstractManipulator
+{
+	/** @var array the array as output by getimagesize */
+	public $sizes = [];
+
+	/** @var int the exif orientation value */
+	public $orientation = 0;
+
+	/** @var string the name of the file we are manipulating */
+	protected $_fileName = '';
+
+	/** @var int width of the image, updated after any manipulation */
+	protected $_width = 0;
+
+	/** @var int the height of the image, updated after any manipulation */
+	protected $_height = 0;
+
+	/**
+	 * AbstractManipulator constructor.
+	 *
+	 * @param $fileName
+	 */
+	abstract public function __construct($fileName);
+
+	/**
+	 * If a given manipulator can be used
+	 *
+	 * @return mixed
+	 */
+	abstract public static function canUse();
+
+	/**
+	 * Resize an image proportionally to fit within the defined max_width and max_height limits
+	 *
+	 * @param int|null $max_width The maximum allowed width
+	 * @param int|null $max_height The maximum allowed height
+	 * @param bool $strip Whether to have IM strip EXIF data as GD will
+	 * @param bool $force_resize = false Whether to override defaults and resize it
+	 */
+	abstract public function resizeImage($max_width = null, $max_height = null, $strip = false, $force_resize = true);
+
+	/**
+	 * Rotate an image based on its EXIF flag, used to correct for smart phone pictures.
+	 *
+	 * @return mixed
+	 */
+	abstract public function autoRotateImage();
+
+	/**
+	 * Creates an image using the supplied text.
+	 *
+	 * @param string $text
+	 * @param int $width
+	 * @param int $height
+	 * @param string $format
+	 *
+	 * @return mixed
+	 */
+	abstract public function generateTextImage($text, $width = 100, $height = 75, $format = 'png');
+
+	/**
+	 * Loads a image file into the image engine for processing
+	 *
+	 * @return bool|mixed
+	 */
+	abstract function createImageFromFile();
```

Comment:
scrutinizer says the scope is missing. :angel: 
Same for other abstract in this class.

Reply:
Yup ... Guilty